### PR TITLE
Add numerical stabilization for difference of exponentials

### LINKED ISSUE #1303: Add numerical stabilization for difference of exponentials
From https://github.com/aesara-devs/aeppl/pull/131

```python
def logdiffexp(a, b):
    """log(exp(a) - exp(b))"""
    return a + at.log1mexp(b - a)
```

### PR DESCRIPTION
**Pull request by Smit-create**
Closes #1303.

### CODE REVIEW

**Comment by Smit-create** [comment]
I'm unable to test the changes locally because of some lock issues in macOS. If I run the following command:
```console
% pytest -v tests/tensor/test_math.py::test_log1mexp_grad_lim
```

<details>
<summary> Timeout Error </summary>
<p>

```console
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <filelock._unix.UnixFileLock object at 0x1368913c0>, timeout = 120
poll_interval = 0.05

    def acquire(
        self,
        timeout: float | None = None,
        poll_interval: float = 0.05,
        *,
        poll_intervall: float | None = None,
        blocking: bool = True,
    ) -> AcquireReturnProxy:
        """
        Try to acquire the file lock.
    
        :param timeout: maximum wait time for acquiring the lock, ``None`` means use the default :attr:`~timeout` is and
         if ``timeout < 0``, there is no timeout and this method will block until the lock could be acquired
        :param poll_interval: interval of trying to acquire the lock file
        :param poll_intervall: deprecated, kept for backwards compatibility, use ``poll_interval`` instead
        :param blocking: defaults to True. If False, function will return immediately if it cannot obtain a lock on the
         first attempt. Otherwise this method will block until the timeout expires or the lock is acquired.
        :raises Timeout: if fails to acquire lock within the timeout period
        :return: a context object that will unlock the file when the context is exited
    
        .. code-block:: python
    
            # You can use this method in the context manager (recommended)
            with lock.acquire():
                pass
    
            # Or use an equivalent try-finally construct:
            lock.acquire()
            try:
                pass
            finally:
                lock.release()
    
        .. versionchanged:: 2.0.0
    
            This method returns now a *proxy* object instead of *self*,
            so that it can be used in a with statement without side effects.
    
        """
        # Use the default timeout, if no timeout is provided.
        if timeout is None:
            timeout = self.timeout
    
        if poll_intervall is not None:
            msg = "use poll_interval instead of poll_intervall"
            warnings.warn(msg, DeprecationWarning, stacklevel=2)
            poll_interval = poll_intervall
    
        # Increment the number right at the beginning. We can still undo it, if something fails.
        with self._thread_lock:
            self._lock_counter += 1
    
        lock_id = id(self)
        lock_filename = self._lock_file
        start_time = time.monotonic()
        try:
            while True:
                with self._thread_lock:
                    if not self.is_locked:
                        _LOGGER.debug("Attempting to acquire lock %s on %s", lock_id, lock_filename)
                        self._acquire()
    
                if self.is_locked:
                    _LOGGER.debug("Lock %s acquired on %s", lock_id, lock_filename)
                    break
                elif blocking is False:
                    _LOGGER.debug("Failed to immediately acquire lock %s on %s", lock_id, lock_filename)
                    raise Timeout(self._lock_file)
                elif 0 <= timeout < time.monotonic() - start_time:
                    _LOGGER.debug("Timeout on acquiring lock %s on %s", lock_id, lock_filename)
>                   raise Timeout(self._lock_file)
E                   filelock._error.Timeout: The file lock 'The file lock '/Users/thebigbool/.aesara/compiledir_macOS-13.0-x86_64-i386-64bit-i386-3.10.8-64/.lock' could not be acquired.
E                   Apply node that caused the error: Elemwise{Composite{Switch(IsInf(Composite{(i0 / expm1((-i1)))}(i0, i1)), i2, Composite{(i0 / expm1((-i1)))}(i0, i1))}}(TensorConstant{-1.0}, x, TensorConstant{-inf})
E                   Toposort index: 0
E                   Inputs types: [TensorType(float64, ()), TensorType(float64, ()), TensorType(float32, ())]
E                   
E                   HINT: Use a linker other than the C linker to print the inputs' shapes and strides.
E                   HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
E                   HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.' could not be acquired.

../../opt/anaconda3/envs/aesara-dev/lib/python3.10/site-packages/filelock/_api.py:183: Timeout
============================= slowest 50 durations =============================
120.03s call     tests/tensor/test_math.py::test_log1mexp_grad_lim

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
=========================== short test summary info ============================
FAILED tests/tensor/test_math.py::test_log1mexp_grad_lim - filelock._error.Timeout: The file lock 'The file lock '/Users/thebigbool/.a...
======================== 1 failed in 140.14s (0:02:20) =========================
```

</p>
</details>

**Comment by Smit-create** [comment]
Seems like some OS issue as the tests already run on CI. Any pointers on this @brandonwillard?

**Comment by brandonwillard** [comment]
> Seems like some OS issue as the tests already run on CI. Any pointers on this @brandonwillard?

I would first check that `~/.aesara`&mdash;or wherever your Aesara cache directory (i.e. `base_compiledir`) is located&mdash;isn't pointing to a remote directory or something similar that could trip up `filelock`.

Does this happen literally every time something is compiled in Aesara, or only when that particular test is run?

File: aesara/tensor/rewriting/math.py [hunk 1]
```diff
@@ -311,6 +311,35 @@ def local_exp_log(fgraph, node):
...
 
 
+@register_specialize
+@node_rewriter([Elemwise])
+def log_diff_exp(fgraph, node):
+    # Case for log(exp(a) - exp(b)) -> a + log1mexp(b - a)
+    x = node.inputs[0]
+
+    if not isinstance(node.op, Elemwise):
+        return
```
Comment (brandonwillard) [inline aesara/tensor/rewriting/math.py:321]:
```suggestion
```

These checks against the `tracks` objects (i.e. `[Elemwise]`) are now automatic in the objects constructed by `node_rewriter`.  Again, we still need to update the old rewrites to reflect this at some point.

File: aesara/tensor/rewriting/math.py [hunk 2]
```diff
@@ -311,6 +311,35 @@ def local_exp_log(fgraph, node):
         return [exp(x)]
 
 
+@register_specialize
+@node_rewriter([Elemwise])
+def log_diff_exp(fgraph, node):
+    # Case for log(exp(a) - exp(b)) -> a + log1mexp(b - a)
```
Comment (brandonwillard) [inline aesara/tensor/rewriting/math.py:317]:
```suggestion
    r"""Rewrite that changes ``log(exp(a) - exp(b))`` to ``a + log1mexp(b - a)``."""
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by Smit-create** [comment]
> Does this happen literally every time something is compiled in Aesara, or only when that particular test is run?

This happens to me whenever I try to run the tests.

File: tests/tensor/test_math.py [hunk 3]
```diff
@@ -3521,3 +3521,16 @@ def test_infer_shape(self):
                 [x1, x2],
                 self.op_class,
             )
+
+
+def test_logdiffexp():
+    x = fmatrix()
+    y = fmatrix()
+    f = function([x, y], log(exp(x) - exp(y)))
```
[the remark this answers is not in the document]
Comment (Smit-create) [reply tests/tensor/test_math.py]: Thanks, understood!

File: aesara/tensor/rewriting/math.py [hunk 4]
```diff
@@ -311,6 +311,35 @@ def local_exp_log(fgraph, node):
         return [exp(x)]
 
 
+@register_specialize
+@node_rewriter([Elemwise])
```
[the remark this answers is not in the document]
Comment (Smit-create) [reply aesara/tensor/rewriting/math.py]: Yeah, makes sense. Thanks!

**Comment by brandonwillard** [comment]
> > Does this happen literally every time something is compiled in Aesara, or only when that particular test is run?
> 
> This happens to me whenever I try to run the tests.

Very interesting!  Don't fix anything just yet; I would like to get on a call first and check it out.  If it's an Aesara bug, this would be a great opportunity to squash it.

In the meantime, you can always test things using the Python backend by setting the Aesara `cxx` setting to an empty string.

**Comment by brandonwillard** [comment]
> > > Does this happen literally every time something is compiled in Aesara, or only when that particular test is run?
> > 
> > 
> > This happens to me whenever I try to run the tests.
> 
> Very interesting! Don't fix anything just yet; I would like to get on a call first and check it out. If it's an Aesara bug, this would be a great opportunity to squash it.
> 
> In the meantime, you can always test things using the Python backend by setting the Aesara `cxx` setting to an empty string.

This was solved by via the `environment-arm.yml` discussed in https://github.com/aesara-devs/aesara/issues/909.

**Review by brandonwillard — APPROVED** [review]
Looks good; we just need to squash before/while merging.

**Review by rlouf — APPROVED** [review]

**Comment by rlouf** [comment]
Looks good, merging!

**Pull request merged**